csrc/ir/nodes.cpp

```diff
@@ -5211,6 +5215,71 @@ bool ForLoop::isGroup() const {
        typeid(kir::GroupedGridWelford)});
 }
 
+namespace {
+
+//! A utility class to check if runtime reduction exists
+class RuntimeReductionFinder : kir::ConstIrVisitor {
+ public:
+  static bool exists(const Expr* expr) {
+    RuntimeReductionFinder finder;
+    finder.handle(std::vector<const Expr*>{expr});
+    return finder.is_found_;
+  }
+
+ private:
+  using kir::ConstIrVisitor::handle;
+
+  // For ReductionOp and WelfordOp, look for block and grid reduction.
+  // For other reductions, runtime function is always called.
+  void dispatch(const Expr* expr) final {
+    // Lambda that extracts the TensorView from an op's output and
+    // returns true if its domain contains a block or grid reduction.
+    auto checkBlockGridReduction = [](const auto* op) -> bool {
+      TensorView* out_tv = nullptr;
+      if (auto tv_idx = dynamic_cast<kir::TensorIndex*>(op->out())) {
+        out_tv = tv_idx->view();
+      } else if (auto tv = dynamic_cast<TensorView*>(op->out())) {
+        out_tv = tv;
+      }
+      if (out_tv) {
+        const auto domain = out_tv->domain();
+        return domain->hasBlockReduction() || domain->hasGridReduction();
+      }
+      return false;
+    };
+
+    if (auto rop = dynamic_cast<const ReductionOp*>(expr)) {
+      if (checkBlockGridReduction(rop)) {
+        is_found_ = true;
+        return;
+      }
+    } else if (auto wop = dynamic_cast<const WelfordOp*>(expr)) {
+      if (checkBlockGridReduction(wop)) {
+        is_found_ = true;
+        return;
+      }
```

Comment:
Why is this necessary? Isn' the below `else if` block just sufficient?

Reply:
The explicit check of block or grid reduction is not necessary as in our previous comment https://github.com/NVIDIA/Fuser/pull/3818#discussion_r1945145489
But I thought based on our offline discussion, we want to do a more explicit check of the use of runtime reduction function. Removed for these unnecessary checks for code clarity.

Reply:
The code here is fine as is, but I just got confused because these two blocks seem to try to find the same pattern exprs. What I suggested was more general pattern matching, i.e., just finding any reduction tensor with block or grid parallelization, no matter what expr is associated with it. Here, both cases only look at `ReductionOp` and `WelfordReductionOp`.